service/listeners.go

```diff
@@ -128,42 +128,33 @@ type readRequest struct {
 
 type virtualPacketConn struct {
 	net.PacketConn
-	mu          sync.Mutex // Mutex to protect access to the channels
 	readCh      chan readRequest
+	closeCh     chan struct{}
 	onCloseFunc OnCloseFunc
 }
 
-func (pc *virtualPacketConn) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
+func (pc *virtualPacketConn) ReadFrom(p []byte) (int, net.Addr, error) {
 	respCh := make(chan struct {
 		n    int
 		addr net.Addr
 		err  error
 	}, 1)
 
-	pc.mu.Lock()
-	if pc.readCh == nil {
-		pc.mu.Unlock()
-		return 0, nil, net.ErrClosed
-	}
-	pc.readCh <- readRequest{
+	select {
+	case pc.readCh <- readRequest{
 		buffer: p,
 		respCh: respCh,
+	}:
+	case <-pc.closeCh:
+		return 0, nil, net.ErrClosed
 	}
-	pc.mu.Unlock()
 
 	resp := <-respCh
 	return resp.n, resp.addr, resp.err
 }
 
 func (pc *virtualPacketConn) Close() error {
-	pc.mu.Lock()
-	defer pc.mu.Unlock()
-
-	if pc.readCh == nil {
-		return nil
-	}
-	pc.readCh = nil
-
+	close(pc.closeCh)
 	if pc.onCloseFunc != nil {
 		onCloseFunc := pc.onCloseFunc
```

Comment:
There's a race condition here if Close is called twice. One fix could be to use an atomic swap with nil.
Perhaps we should add a comment that Close must be called once, and only once per virtualPacketConn.

Reply:
Ah you're right. I forgot the mutex wasn't just used for the `readCh`. Let me bring it back.